python/ray/workflow/http_event_provider.py

```diff
@@ -0,0 +1,240 @@
+
+        @app.post("/arrivals")
+        async def create_event_payload(self, req: Request):
+
+        async def get_event_payload(self, workflow_id, event_key):
+
+        async def report_checkpointed(self, workflow_id, event_key, confirmation):
+
+    ```
+
+    Example Usage
+    =============
+    >>> from ray.workflow.http_event_provider import HTTPEventProvider, HTTPListener
+    >>> ray.init(address='auto', namespace='serve')
+    >>> serve.start(detached=True)
+    >>> HTTPEventProvider.deploy()
+    >>> event_step = workflow.wait_for_event( # doctest: +SKIP
+    ...     HTTPListener, workflow_id='', event_key='')
+    >>> handle_event = ... # doctest: +SKIP
+    >>> handle_event.step(event_step).run() # doctest: +SKIP
+    >>>
+    >>> On a separate python process, it sends an event to the HTTPEventProvider.
+    >>> import requests
+    >>> resp = requests.post('http://127.0.0.1:8000/event/arrivals',
+    ...     json={'workflow_id':'my_wf','event_key':'my_key','event_payload':'testMessage'})
+    >>>
+
+    """
+
+    def __init__(self, event_coordinator_ack: Optional[bool] = False):
+        import nest_asyncio
+        nest_asyncio.apply()
+        self.event_key_payload: Dict[str, Dict[Any, Any]] = {}
+        self.event_coordinator_ack = event_coordinator_ack
+        if self.event_coordinator_ack:
+            self.event_key_checkpointed: Dict[str, Dict[Any, Any]] = {}
+        else:
+            self.payload_record: Dict[Any] = {}
+
+    @app.post("/arrivals")
+    async def create_event_payload(self, req: Request):
+        req_json = await req.json()
+        workflow_id = req_json['workflow_id']
+        event_key = req_json['event_key']
+        event_payload = req_json['event_payload']
+        if workflow_id not in self.event_key_payload.keys():
+            return f"Event Dropped: not registered, please resend."
+        if event_key not in self.event_key_payload[workflow_id].keys():
+            return f"Event Dropped: not registered, please resend."
+        self.event_key_payload[workflow_id][event_key].set_result(event_payload)
+
+        if self.event_coordinator_ack:
+            # wait for event listener to report event has been checkpointed before acknowledging
+            if workflow_id not in self.event_key_checkpointed:
+                self.event_key_checkpointed[workflow_id] = {}
+            self.event_key_checkpointed[workflow_id][event_key] = asyncio.Future()
+            # await until confirmation of checkpointed is received.
+            confirmed = await self.event_key_checkpointed[workflow_id][event_key]
+            self.event_key_checkpointed[workflow_id].pop(event_key)
+            if confirmed:
+                return f"ACK_OK: event received and checkpointed successfully."
+            else:
+                # the ECA confirms the failure of checkpointing the event.
+                return f"ERROR: event received but failed in checkpointing."
```

Comment:
please send the right resp (https://fastapi.tiangolo.com/tutorial/handling-errors/) when error happened